sale_delivery_state/models/sale_order.py

```diff
@@ -0,0 +1,114 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 Akretion (http://www.akretion.com).
+# @author Pierrick BRUN <pierrick.brun@akretion.com>
+# Copyright 2018 Camptocamp
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, fields, api
+from odoo.tools import float_is_zero, float_compare
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    delivery_state = fields.Selection([
+        ('no', 'No delivery'),
+        ('unprocessed', 'Unprocessed'),
+        ('partially', 'Partially processed'),
+        ('done', 'Done')],
+        string='Delivery state',
+        compute='_compute_delivery_state',
+        store=True
+    )
+
+    force_delivery_state = fields.Boolean(
+        string='Force delivery state',
+        help=("Allow to enforce done state of delivery, for instance if some"
+              " quantities were cancelled")
+    )
+
+    def _all_qty_delivered(self):
+        """
+        Returns True if all line have qty_delivered >= to ordered quantities
+
+        If `delivery` module is installed, ignores the lines with delivery costs
+
+        :returns: boolean
+        """
+        self.ensure_one()
+        # Skip delivery costs lines
+        sale_lines = self.order_line
+        Carrier = self.env.get('delivery.carrier')
+        if Carrier:
+            sale_lines = sale_lines.filtered(
+                lambda rec: not rec.is_delivery_cost())
```

Comment:
@yvaucher If delivery is installed, a new field is added on the line to specify is a line is a delivery product or not: `is_delivery`
https://github.com/odoo/odoo/blob/12.0/addons/delivery/models/sale_order.py#L130

IMO we should remove the `is_delivery_cost`method and rely on this attribute. That should improve the performances by avoiding an sql search on each call to is_delivery_cost.

Reply:
@lmignon good catch, I missed that field. Will do the change.

Reply:
Ping @yvaucher